File: engine/compile/ops/transpose.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
#
# Copyright (c) 2021 Intel Corporation
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from .op import Operator, operator_registry
from .tensor import Tensor
from ..graph_utils import list2str


# tf.transpose(a, perm=None, conjugate=False, name='transpose')
# The returned tensor's dimension i will correspond to the input dimension perm[i].
# If perm is not given, it is set to (n-1...0), where n is the rank of the input tensor.
# Hence by default, this operation performs a regular matrix transpose on 2-D input Tensors.
# If conjugate is True and a.dtype is either complex64 or complex128 then the values of a are
# conjugated and transposed.
# normally, we don't need the param 'conjugate'
@operator_registry(operator_type='Transpose')
class Transpose(Operator):
    def __init__(self):
        super().__init__()

    def set_attr(self, framework, node):

        if framework == "tensorflow":
            if self._input_tensors[1].source_op == []:
                dst_perm = list(self._input_tensors[1].data)
                self._attr['dst_perm'] = list2str(dst_perm)
                src_perm = [i for i in range(len(dst_perm))]
                self._attr['src_perm'] = list2str(src_perm)
                _ = self._input_tensors.pop()

        if framework == "onnxruntime":
            dst_perm = node.attribute[0].ints
            src_perm = [i for i in range(len(dst_perm))]
            self._attr['src_perm'] = list2str(src_perm)
            self._attr['dst_perm'] = list2str(dst_perm)

File: engine/executor/src/operators/transpose.cpp
//  Copyright (c) 2021 Intel Corporation
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//    http://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

#include "transpose.hpp"

namespace executor {

TransposeOperator::TransposeOperator(const OperatorConfig& conf) : Operator(conf) {
  auto attrs_map = operator_conf_.attributes();

  auto iter = attrs_map.find("src_perm");
  if (iter != attrs_map.end()) {
    StringSplit<int64_t>(&src_perm_, attrs_map["src_perm"], ",");
  }
  iter = attrs_map.find("dst_perm");
  if (iter != attrs_map.end()) {
    StringSplit<int64_t>(&dst_perm_, attrs_map["dst_perm"], ",");
  }
  iter = attrs_map.find("output_dtype");
  if (iter != attrs_map.end()) {
    output_dtype_ = attrs_map["output_dtype"];
  }
}

TransposeOperator::~TransposeOperator() {}

void TransposeOperator::Reshape(const vector<Tensor*>& input, const vector<Tensor*>& output) {
  //// Part1: Derive operator's user proper shape and strides
  // 1.1: Prepare Tensor origin shape
  vector<int64_t> src_shape = input[0]->shape();

  // 1.2 Get tensor's adjusted shapes
  // dst shape
  vector<int64_t> dst_shape_origin = src_shape;
  vector<int64_t> dst_shape = GetShapes(dst_shape_origin, dst_perm_);

  // 1.5 Set dst shape and strides
  output[0]->set_shape(dst_shape);
  output[0]->set_dtype(output_dtype_);

  for (size_t i = 0; i < 4; ++i) {
    perm_[i] = dst_perm_[i];
    src_shape_[i] = static_cast<Eigen::DenseIndex>(input[0]->shape()[i]);
    dst_shape_[i] = static_cast<Eigen::DenseIndex>(output[0]->shape()[i]);
  }
}

// 2. inference kernel(for int8 and f32)
void TransposeOperator::Forward(const vector<Tensor*>& input, const vector<Tensor*>& output) {
  auto src_data = input[0]->data();
  auto dst_data = output[0]->mutable_data();
  string dtype = output[0]->dtype();
  if (dtype == "fp32") {
    auto input = Eigen::TensorMap<Eigen::Tensor<const float, 4>>(reinterpret_cast<const float*>(src_data), src_shape_);
    auto output = Eigen::TensorMap<Eigen::Tensor<float, 4>>(reinterpret_cast<float*>(dst_data), dst_shape_);
    output = input.shuffle(perm_);
  } else if (dtype == "s8") {
    auto input = Eigen::TensorMap<Eigen::Tensor<const char, 4>>(reinterpret_cast<const char*>(src_data), src_shape_);
    auto output = Eigen::TensorMap<Eigen::Tensor<char, 4>>(reinterpret_cast<char*>(dst_data), dst_shape_);
    output = input.shuffle(perm_);
  }

  this->unref_tensors(input);
}

REGISTER_OPERATOR_CLASS(Transpose);
}  // namespace executor
